lib/linked_list.rb

```diff
@@ -18,79 +18,150 @@ def initialize
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(value)
-      raise NotImplementedError
+      new_node = Node.new(value, @head)
+      @head = new_node
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(value)
-      raise NotImplementedError
+      current = @head
+      until current.nil?
+        if current.data == value
+          return true
+        else
+          current = current.next
+        end
+      end
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_max
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+      max_value = current.data
+      until current.nil?
+        if current.data > max_value
+          max_value = current.data
+        end
+          current = current.next
+      end
+      return max_value
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_min
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+      min_value = current.data
+      until current.nil?
+        if current.data < min_value
+          min_value = current.data
+        end
+          current = current.next
+      end
+      return min_value
     end
 
 
     # Additional Exercises 
     # returns the value in the first node
     # returns nil if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first
-      raise NotImplementedError
+      return nil if @head.nil?
+      return @head.data
     end
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(value)
-      raise NotImplementedError
+      current = @head
+      until current.nil?
+        current = current.next
+      end
+      current = Node.new(value)

```

Comment:
This was not linking in the new node to the list. 

```suggestion
      if @head.nil?
        self.add_first(value)
      end
      
      until current.next.nil?
        current = current.next
      end
      current.next = Node.new(value)

```